sound/soc/sof/sof-audio.c

```diff
@@ -256,29 +256,133 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback)
+ */
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int err;
+	int ret = 0;
+
+	/* free all widgets even in case of error to keep use counts balanced */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
```

Comment:
how do we know that p->walking was properly reset? defensive programming would be nice with this field explicitly cleared before starting the recursion - maybe with a helper doing the recursion after clearing the field?

Reply:
I think it is a bit of a n overkill, Pierre. With p->walking set to false after every successful/failed walk, I dont think we need explicit clearing before.

Reply:
ok, I hope I won't have to say "I told you so" :-)